glottolog3/views.py

```diff
@@ -398,41 +399,144 @@ def bp_api_search(request):
 
     languoid = DBSession.query(Language) \
                         .filter_by(id='{0}'.format(gcode)) \
                         .first()
 
-    identifier = Identifier(
-        (name, type, desc, lang),
-        id='{0}-{1}-{2}-{3}'.format(
-        slug(name), slug(type), slug(desc or ''), lang),
-        name=name,
-        type=type,
-        description=desc,
-        lang=lang)
-
+    if languoid == None:
+        request.response.status = 404 
+        return {'error': 'No language found with glottocode: {}'.format(gcode)}
+    
+    try:
+        identifier, errors = IdentifierSchema().load(request.json_body)
+    except (ValueError, ValidationError) as e:
+        request.response.status = 400
+        return {'error': '{}'.format(e)}
+    if errors:
+        request.response.status = 400
+        return {'error': errors}
+    
     try:
         DBSession.add(identifier)
         DBSession.add(
             LanguageIdentifier(language=languoid, identifier=identifier))
         DBSession.flush()
     except exc.SQLAlchemyError as e:
+        request.response.status = 400
         DBSession.rollback()
         return { 'error': '{}'.format(e) }
+    
+    # So we can use identifier object without session
+    DBSession.expunge(identifier)
+    transaction.commit()
+
+    return json.dumps(IdentifierSchema().dump(identifier))
+
+
+def query_identifier(type, name):
+    type = type.lower()
+    name = name.title() if type == 'name' else name.lower()
+
+    id_query = DBSession.query(Identifier) \
+                          .filter(and_(Identifier.name == name,
+                                       Identifier.type == type)) \
+
+    if id_query.count() == 0:
+        raise Exception(('No identifier found with '
+                         'type: {0}, and name: {1}').format(type, name))
+
+    return id_query
+
+
+@view_config(
+        route_name='glottolog.get_identifier',
+        renderer='json')
+def get_identifier(request):
+    try: 
+        identifier = query_identifier(request.matchdict['type'],
+                                      request.matchdict['name']).first()
+    except Exception as e:
+        request.response.status = 404
+        return {'error': str(e)}
+
+    return json.dumps(IdentifierSchema().dump(identifier)) 
+
+
+@view_config(
+        route_name='glottolog.put_identifier',
+        renderer='json')
+def put_identifier(request):
+    ID_REQ_FIELDS = ['name', 'type', 'description', 'lang'] 
+    new_identifier = request.json_body
+    try: 
+        identifier = query_identifier(request.matchdict['type'],
+                                      request.matchdict['name']).first()
+    except Exception as e:
+        request.response.status = 404
+        return {'error': str(e)}
+
+    # Fill missing
+    for field in ID_REQ_FIELDS:
+        if not field in new_identifier:
+            new_identifier[field] = getattr(identifier, field)
+
+    try:
+        data, errors = IdentifierSchema().load(new_identifier)
```

Comment:
you need a partial=True in case someone only wants to update one field, say description

Reply:
I filled in the remaining manually above, I had some trouble with validation when it was a partial (ie. only having name field in body though type field is needed for validating if it's valid for a type)

Reply:
http://marshmallow.readthedocs.io/en/latest/quickstart.html#partial-loading allows you to leave the required fields required and only specify what's allowed to be missing. this works as well, but might as well not write extra code if we can help it

Reply:
I'm not sure we're on the same page -  I also omitted the partial flag because I basically always need the type field and name field to check if the update is valid (regardless of whether or not the user passed it in). My thought was, instead of having separate checks, I can just do it for all the fields that might've been missing, that way I can also directly use the deserialized object from load.